File: TestAsm/src/main/java/com/example/testacm/RSymbols.java
///*
// * Copyright (c) 2017 Yrom Wang
// *
// * Licensed under the Apache License, Version 2.0 (the "License");
// * you may not use this file except in compliance with the License.
// * You may obtain a copy of the License at
// *
// *    http://www.apache.org/licenses/LICENSE-2.0
// *
// * Unless required by applicable law or agreed to in writing, software
// * distributed under the License is distributed on an "AS IS" BASIS,
// * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// * See the License for the specific language governing permissions and
// * limitations under the License.
// */
//
//package com.example.testacm;
//
//import org.objectweb.asm.ClassReader;
//import org.objectweb.asm.ClassVisitor;
//import org.objectweb.asm.FieldVisitor;
//import org.objectweb.asm.MethodVisitor;
//import org.objectweb.asm.Opcodes;
//
//import java.io.IOException;
//import java.io.UncheckedIOException;
//import java.nio.file.Files;
//import java.nio.file.Path;
//import java.util.Arrays;
//import java.util.Collections;
//import java.util.HashMap;
//import java.util.LinkedList;
//import java.util.Map;
//
//import static org.objectweb.asm.ClassReader.SKIP_DEBUG;
//import static org.objectweb.asm.ClassReader.SKIP_FRAMES;
//
///**
// * @author yrom
// */
//class RSymbols {
//    /**
//     * default package!
//     */
//    static final String R_STYLEABLES_CLASS_NAME = "R$styleable";
//
//    // 比如R.string.a=0xd7060001，保存在map里应为为["string.a"] = 0xd7060001
//    private Map<String, Integer> symbols = Collections.emptyMap();
//
//    // 比如R.styleables.b = int[] {1, 2}，保存在map里应该为[styleables.b] = int[] {1, 2}
//    private Map<String, int[]> styleables = new HashMap<>();
//
//    public Integer get(String key) {
//        return symbols.get(key);
//    }
//
//    public boolean containsKey(String key) {
//        return symbols.containsKey(key);
//    }
//
//    public boolean isEmpty() {
//        return symbols.isEmpty() && styleables.isEmpty();
//    }
//
//    public Map<String, int[]> getStyleables() {
//        return Collections.unmodifiableMap(styleables);
//    }
//
//
//    {
//        styleables.put("styleables.a", new int[] {1,2,3});
//        styleables.put("styleables.b", new int[] {1,2,3});
//        styleables.put("styleables.c", new int[] {1,2,3});
//        styleables.put("styleables.d", new int[] {1,2,3});
//    }
//
////    public RSymbols from(Collection<TransformInput> inputs) {
////
////
////        // 找到所有R相关的类
////        final PathMatcher rClassMatcher = FileSystems.getDefault().getPathMatcher("glob:R$*.class");
////
////        final List<Path> paths = inputs.stream()
////                // 所有依赖的包都会在主工程下生成一个对应的R文件
////                .map(TransformInput::getDirectoryInputs)
////
////
////                .flatMap(Collection::stream)
////                .map(this::toStream)
////                .reduce(Stream.empty(), Stream::concat)
////                .collect(Collectors.toList());
////
////        System.out.println("paths:" + paths);
////
////        Stream<Path> stream;
////        if (paths.size() >= Runtime.getRuntime().availableProcessors() * 3) {
////            // use parallel here!
////            stream = paths.parallelStream();
////            symbols = Maps.newConcurrentMap();
////        } else {
////            stream = paths.stream();
////            symbols = Maps.newHashMap();
////        }
////        stream.filter(path -> rClassMatcher.matches(path.getFileName()))
////                .forEach(this::drainSymbols);
////        return this;
////    }
//
//    private void drainSymbols(Path file) {
//        final String filename = file.getFileName().toString();
//        final String typeName = filename.substring(0, filename.length() - ".class".length());
//
//        byte[] bytes;
//        try {
//            bytes = Files.readAllBytes(file);
//        } catch (IOException e) {
//            throw new UncheckedIOException(e);
//        }
//
//
//        ClassVisitor visitor = new ClassVisitor(Opcodes.ASM5) {
//            @Override
//            public FieldVisitor visitField(int access, String name, String desc, String signature, Object value) {
//                // read constant value
//                if (value instanceof Integer) {
//                    String key = typeName + '.' + name;
//                    Integer old = symbols.get(key);
//                    if (old != null && !old.equals(value)) {
//                        throw new IllegalStateException("Value of " + key + " mismatched! "
//                                + "Excepted 0x" + Integer.toHexString(old)
//                                + " but was 0x" + Integer.toHexString((Integer) value));
//                    } else {
//                        symbols.put(key, (Integer) value);
//                    }
//                }
//                return null;
//            }
//
//            @Override
//            public MethodVisitor visitMethod(int access, String name, String desc, String signature, String[] exceptions) {
//                if (access == Opcodes.ACC_STATIC && "<clinit>".equals(name)) {
//
//                    return new MethodVisitor(Opcodes.ASM5) {
//                        int[] current = null;
//                        LinkedList<Integer> intStack = new LinkedList<>();
//
//                        @Override
//                        public void visitIntInsn(int opcode, int operand) {
//                            if (opcode == Opcodes.NEWARRAY && operand == Opcodes.T_INT) {
//                                current = new int[intStack.pop()];
//                            } else if (opcode == Opcodes.BIPUSH) {
//                                intStack.push(operand);
//                            }
//                        }
//
//                        @Override
//                        public void visitLdcInsn(Object cst) {
//                            if (cst instanceof Integer) {
//                                intStack.push((Integer) cst);
//                            }
//                        }
//
//                        @Override
//                        public void visitInsn(int opcode) {
//                            if (opcode >= Opcodes.ICONST_0 && opcode <= Opcodes.ICONST_5) {
//                                intStack.push(opcode - Opcodes.ICONST_0);
//                            } else if (opcode == Opcodes.IASTORE) {
//                                int value = intStack.pop();
//                                int index = intStack.pop();
//                                current[index] = value;
//                            }
//                        }
//
//                        @Override
//                        public void visitFieldInsn(int opcode, String owner, String name, String desc) {
//                            if (opcode == Opcodes.PUTSTATIC) {
//                                int[] old = styleables.get(name);
//                                if (old != null && old.length != current.length && !Arrays.equals(old, current)) {
//                                    throw new IllegalStateException("Value of styleable." + name + " mismatched! "
//                                            + "Excepted " + Arrays.toString(old)
//                                            + " but was " + Arrays.toString(current));
//                                } else {
//                                    styleables.put(name, current);
//                                }
//                                current = null;
//                                intStack.clear();
//                            }
//                        }
//                    };
//                }
//                return null;
//            }
//        };
//
//        new ClassReader(bytes).accept(visitor, SKIP_DEBUG | SKIP_FRAMES);
//    }
//
////    private Stream<Path> toStream(DirectoryInput dir) {
////        try {
////            return Files.walk(dir.getFile().toPath()).filter(Files::isRegularFile);
////        } catch (IOException e) {
////            throw new UncheckedIOException(e);
////        }
////    }
//}

File: TestAcm/src/main/java/com/example/testacm/RSymbols.kt
package com.example.testacm


import org.objectweb.asm.ClassReader
import org.objectweb.asm.ClassVisitor
import org.objectweb.asm.FieldVisitor
import org.objectweb.asm.MethodVisitor
import org.objectweb.asm.Opcodes

import java.io.IOException
import java.io.UncheckedIOException
import java.nio.file.FileSystems
import java.nio.file.Files
import java.nio.file.Path
import java.util.Arrays
import java.util.Collections
import java.util.LinkedList
import java.util.stream.Collectors
import java.util.stream.Stream

import org.objectweb.asm.ClassReader.SKIP_DEBUG
import org.objectweb.asm.ClassReader.SKIP_FRAMES
import java.io.File

/**
 * @author yrom
 */
class RSymbols() {

    companion object {
        /**
         * default package!
         */
        const val R_STYLEABLES_CLASS_NAME = "R\$styleable"


//        /**
//         * 从inputs中解析出所有的R$xxx类
//         */
//        fun collectAllRFiles(inputs: Collection<TransformInput>): RSymbols {
//            val symbols = RSymbols()
//
//            val paths = inputs.stream()
//                    // 所有依赖的包都会在主工程下生成一个对应的R文件
//                    .map {
//                        it.directoryInputs
//                    }
//                    //
//                    .flatMap<DirectoryInput> {
//                         it.stream()
//                    }
//                    .map<Stream<Path>> {
//                        toStream(it)
//                    }
//                    .reduce(Stream.empty()) { stream, stream1 ->
//                        Stream.concat(stream, stream1)
//                    }
//                    .collect(Collectors.toList())
//
//
//            val stream: Stream<Path>
//            if (paths.size >= Runtime.getRuntime().availableProcessors() * 3) {
//                // use parallel here!
//
//                stream = paths.parallelStream()
//                symbols.symbols = Maps.newConcurrentMap()
//            } else {
//                stream = paths.stream()
//                symbols.symbols = Maps.newHashMap()
//            }
//
//            // 找到所有R相关的类
//            val rClassMatcher = FileSystems.getDefault().getPathMatcher("glob:R$*.class")
//            stream.filter {
//                rClassMatcher.matches(it.fileName)
//            } .forEach {
//                drainRSymbolsFromRFile(it, symbols)
//            }
//
//            return symbols
//        }


        /**
         * 从inputs中解析出所有的R$xxx类
         */
        fun collectAllRFiles(inputs: File): RSymbols {
            val symbols = RSymbols()

//
//            val paths = inputs.stream()
//                    // 所有依赖的包都会在主工程下生成一个对应的R文件
//                    .map {
//                        it.directoryInputs
//                    }
//                    //
//                    .flatMap<DirectoryInput> {
//                        it.stream()
//                    }
//                    .map<Stream<Path>> {
//                        toStream(it)
//                    }
//                    .reduce(Stream.empty()) { stream, stream1 ->
//                        Stream.concat(stream, stream1)
//                    }
//                    .collect(Collectors.toList())


            val stream: Stream<Path> = toStream(inputs)
//            println("stream:$stream")
//            if (paths.size >= Runtime.getRuntime().availableProcessors() * 3) {
//                // use parallel here!
//
//                stream = paths.parallelStream()
//                symbols.symbols = Maps.newConcurrentMap()
//            } else {
//                stream = paths.stream()
//                symbols.symbols = Maps.newHashMap()
//            }

            // 找到所有R相关的类
            val rClassMatcher = FileSystems.getDefault().getPathMatcher("glob:R$*.class")
            stream.filter {
                rClassMatcher.matches(it.fileName)
            } .forEach {

                drainRSymbolsFromRFile(it, symbols)
            }

            return symbols
        }


        private fun toStream(file: File): Stream<Path> {
//            println("toStream:${file.exists()}")
            try {
                return Files.walk(file.toPath()).filter {path ->
//                    println("filter:$path")
                    Files.isRegularFile(path)
                }
            } catch (e: IOException) {
                throw UncheckedIOException(e)
            }
        }




        private fun drainRSymbolsFromRFile(file: Path, symbols: RSymbols) {

            println("drainRSymbolsFromRFile:$file")

            val filename = file.fileName.toString()
            val typeName = filename.substring(0, filename.length - ".class".length)
            val bytes: ByteArray

            try {
                bytes = Files.readAllBytes(file)
            } catch (e: IOException) {
                throw UncheckedIOException(e)
            }


            val visitor = object : ClassVisitor(Opcodes.ASM5) {

                // 收集非R$.styleables的R$xxx类
                override fun visitField(access: Int, name: String?, desc: String?, signature: String?, value: Any?): FieldVisitor? {
                    // read constant value
                    if (value !is Int) return null // 如果不是int值，比表示

                    val key = "$typeName.$name"
                    val old = symbols.symbols[key]

                    if (old != null && old != value) {
                        throw IllegalStateException("Value of " + key + " mismatched! "
                                + "Excepted 0x" + Integer.toHexString(old)
                                + " but was 0x" + Integer.toHexString(value))
                    } else {
                        symbols.symbols[key] = value
                    }
                    return null
                }

                override fun visitMethod(access: Int, name: String?, desc: String?, signature: String?, exceptions: Array<String>?): MethodVisitor? {

                    return if (access == Opcodes.ACC_STATIC && "<clinit>" == name) {

                        // public static final int[] a = new int[] {100, 200, 300}
//        NEWARRAY T_INT  创建int数组
//        DUP
//        ICONST_0  将100放入数组index 0
//        BIPUSH 100
//        IASTORE
//                DUP
//        ICONST_1 将200放入数组index 1
//        SIPUSH 200
//        IASTORE
//                DUP
//        ICONST_2 将300放入数组index 2
//        SIPUSH 300
//        IASTORE
//        PUTSTATIC com/example/testacm/TestA.a : [I //把数组赋值给类TestA的静态域a


                        object : MethodVisitor(Opcodes.ASM5) {

                            var current: IntArray? = null
                            var intStack = LinkedList<Int>()

                            override fun visitIntInsn(opcode: Int, operand: Int) {
                                if (opcode == Opcodes.NEWARRAY && operand == Opcodes.T_INT) {
                                    current = IntArray(intStack.pop())
                                } else if (opcode == Opcodes.BIPUSH || opcode == Opcodes.SIPUSH) {
                                    intStack.push(operand)
                                }
                            }

                            override fun visitLdcInsn(cst: Any) {
                                if (cst is Int) {
                                    intStack.push(cst)
                                }
                            }

                            override fun visitInsn(opcode: Int) {
                                if (opcode >= Opcodes.ICONST_0 && opcode <= Opcodes.ICONST_5) {
                                    intStack.push(opcode - Opcodes.ICONST_0)
                                } else if (opcode == Opcodes.IASTORE) {
                                    val value = intStack.pop()
                                    val index = intStack.pop()
                                    current?.set(index, value)
                                }
                            }

                            override fun visitFieldInsn(opcode: Int, owner: String, name: String, desc: String) {
                                if (opcode == Opcodes.PUTSTATIC) {
                                    val old = symbols.styleables[name]
                                    if (old != null && old.size != current!!.size && !Arrays.equals(old, current)) {
                                        throw IllegalStateException("Value of styleable." + name + " mismatched! "
                                                + "Excepted " + Arrays.toString(old)
                                                + " but was " + Arrays.toString(current))
                                    } else {
                                        symbols.styleables[name] = current!!
                                    }
                                    current = null
                                    intStack.clear()
                                }
                            }
                        }
                    } else null
                }
            }

            ClassReader(bytes).accept(visitor, SKIP_DEBUG or SKIP_FRAMES)
        }


    }

    /**
     * 比如R.string.a=0xd7060001，保存在map里应为为["string.a"] = 0xd7060001
     */
    private var symbols: MutableMap<String, Int> = mutableMapOf()

    /**
     * 比如R.styleables.b = int[] {1, 2}，保存在map里应该为[styleables.b] = int[] {1, 2}
     */
    val styleables: MutableMap<String, IntArray> = mutableMapOf()

    val isEmpty: Boolean
        get() = symbols.isEmpty() && styleables.isEmpty()

    operator fun get(key: String): Int? {
        return symbols[key]
    }

    fun containsKey(key: String): Boolean {
        return symbols.containsKey(key)
    }

//    fun getStyleables(): Map<String, IntArray> {
//        return Collections.unmodifiableMap(styleables)
//    }
}
